File: ValidSudoko.java

import java.util.HashSet;
import java.util.Set;

public class ValidSudoko {
    public boolean isValidSudoku(char[][] board) {
        Set<String> set = new HashSet<>();
        String row = "row";
        String col = "col";
        String square = "square";
        String valueSeparator = "|";
        for(int i=0; i<board.length; i++){
            for(int j=0; j<board.length; j++){
                if(board[i][j] == '.')
                    continue;

                String rowVal = row + i + valueSeparator + board[i][j];
                if(!set.add(rowVal))
                    return false;
            
                String colVal = col + j + valueSeparator + board[i][j];
                if(!set.add(colVal))
                    return false;
                String squareVal = square + i/3 + "" +j/3 + valueSeparator + board[i][j];
                if(!set.add(squareVal))
                    return false;
                
            }
        }
        return true;     
    }

    public static void main(String[] args) {
        char [][]board = new char[][] {
         {'5','3','.','.','7','.','.','.','.'}
        ,{'6','.','.','1','9','5','.','.','.'}
        ,{'.','9','8','.','.','.','.','6','.'}
        ,{'8','.','.','.','6','.','.','.','3'}
        ,{'4','.','.','8','.','3','.','.','1'}
        ,{'7','.','.','.','2','.','.','.','6'}
        ,{'.','6','.','.','.','.','2','8','.'}
        ,{'.','.','.','4','1','9','.','.','5'}
        ,{'.','.','.','.','8','.','.','7','9'}};

        ValidSudoko sudokoSolver = new ValidSudoko();
        boolean isValid = sudokoSolver.isValidSudoku(board);
        System.out.println(isValid);
        
    }
}

File: validSudoko.go
package main

import "fmt"

func isValidSudoku(board [][]byte) bool {
	size := 9

	for row := 0; row < size; row++ {
		rowMap := make(map[byte]bool)
		colMap := make(map[byte]bool)
		boxMap := make(map[byte]bool)

		for col := 0; col < size; col++ {
			if board[row][col] != '.' {
				if rowMap[board[row][col]] {
					return false
				}
				rowMap[board[row][col]] = true
			}
			if board[col][row] != '.' {
				if colMap[board[col][row]] {
					return false
				}
				colMap[board[col][row]] = true
			}
			boxRow := (row/3)*3 + col/3
			boxCol := (row%3)*3 + col%3
			if board[boxRow][boxCol] != '.' {
				if boxMap[board[boxRow][boxCol]] {
					return false
				}
				boxMap[board[boxRow][boxCol]] = true
			}
		}
	}
	return true
}

func main() {
	board := [][]byte{{'.', '.', '.', '.', '5', '.', '.', '1', '.'}, {'.', '4', '.', '3', '.', '.', '.', '.', '.'}, {'.', '.', '.', '.', '.', '3', '.', '.', '1'}, {'8', '.', '.', '.', '.', '.', '.', '2', '.'}, {'.', '.', '2', '.', '7', '.', '.', '.', '.'}, {'.', '1', '5', '.', '.', '.', '.', '.', '.'}, {'.', '.', '.', '.', '.', '2', '.', '.', '.'}, {'.', '2', '.', '9', '.', '.', '.', '.', '.'}, {'.', '.', '4', '.', '.', '.', '.', '.', '.'}}
	k := isValidSudoku(board)
	fmt.Print(k)
}
